### PR TITLE
Plugins API v2: Loader, Base API, and Test Harness

### PR DESCRIPTION
**Pull request by clintoncwolfe**
First bits of the v2 plugins API.
Includes:
 * Move v1 plugin core to a new location under lib/inspec/plugins/v1
 * v2 plugin registry
 * v2 plugin loader
 * v2 plugin config file system
 * test harness for unit and functional tests
 * v2 base plugin API and core DSL
 * v2 plugin subtype, `mock_plugin_type`, for testing

All functional tests pass for me locally; unit tests usually pass, but I do see one sporadic failure, likely due to test ordering; hoping to pin that down.

UPDATE: At this point, I'm interested in any and all feedback - thanks!

UPDATE: Travis run [8251.2](https://travis-ci.org/inspec/inspec/jobs/413154684) is an example of the sporadically failing unit test.  I have added a skip for now, along with an explanation in the test code explaining why I think a skip is the right answer for now.

### CODE REVIEW

File: docs/dev/plugins.md
Comment (jquick) [inline docs/dev/plugins.md]: Should we use the current `~/.inspec/` folder since we already have it? I feel like adding another folder just for a few configs is maybe overkill. I am a fan of the usual `config` folder inside or even just putting them in the root if its only one or two for the user experience.

File: lib/inspec/plugin/v2.rb
Comment (jquick) [inline lib/inspec/plugin/v2.rb]: Should this grab the base `Inspec::Error` instead?

File: lib/inspec/plugin/v2/loader.rb
Comment (jquick) [inline lib/inspec/plugin/v2/loader.rb]: Should we use the `Inspec::Log.error` for this stuff?

**Review by jquick — COMMENTED** [review]
This is pretty amazing stuff @clintoncwolfe !

A few things I loved:
The organization under `lib/inspec/plugin/v?`.
Making plugin types extendable to new ones in the future (very generic implementation)
Your Registry plugin

I don't have any issues with the high level choices on this PR and I think its a really good foundation.

File: lib/inspec/plugin/v2/loader.rb
Comment (clintoncwolfe) [reply lib/inspec/plugin/v2/loader.rb]: Possibly.  We have a might have a load order issue - the plugin loader executes *very* early, and Inspec::Log might not be available yet (which, uh, we should fix, if so).

Very good idea, I'll look into it.

File: docs/dev/plugins.md
Comment (clintoncwolfe) [reply docs/dev/plugins.md]: Sorry, the doc there is from like day 1, and is quite stale.  I agree, I hit that early on, and changed to use plain ~/.inspec - see https://github.com/inspec/inspec/blob/cw/plugins-loader/lib/inspec/plugin/v2/loader.rb#L125

Also uses JSON instead of YAML, since we had precedent.

File: lib/bundles/inspec-init/cli.rb
Comment (chris-rock) [inline lib/bundles/inspec-init/cli.rb:5]: Do we envision to refactor those built-in plugins to use V2?

File: test/unit/mock/config_dirs/basic/plugins.yaml
Comment (chris-rock) [inline test/unit/mock/config_dirs/basic/plugins.yaml]: Have we documented the yml behaviour? I may have missed it.

File: lib/inspec/plugin/v2/status.rb
Comment (chris-rock) [inline lib/inspec/plugin/v2/status.rb:10]: Yeahhhh

**Review by chris-rock — COMMENTED** [review]
Awesome work @clintoncwolfe

File: test/unit/mock/config_dirs/basic/plugins.yaml
Comment (clintoncwolfe) [reply test/unit/mock/config_dirs/basic/plugins.yaml]: Ahh, that file is cruft.  It's using JSON now, and is documented here - https://github.com/inspec/inspec/pull/3278/files#diff-1aac973df988b8de466d788e758d6303R35

File: lib/bundles/inspec-init/cli.rb
Comment (clintoncwolfe) [reply lib/bundles/inspec-init/cli.rb:5]: Yes, as a stretch goal.  Things that are already bundled - like inspec-habitat - are easy to do, and will gain a lot from load-on-demand.  We can also push things like inspec exec, inspec artifact, etc into CLI plugins, which I think would be great for reducing complexity and making testing easier; but I haven't looked into it in any detail to search for obstacles.  That refactoring could be done post-3.0.

**Review by jquick — APPROVED** [review]
Thanks @clintoncwolfe !

File: test/unit/plugin/v2/loader_test.rb
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/loader_test.rb]: Could we remove these commented tests?

File: test/unit/plugin/v2/installer_test.rb
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/installer_test.rb]: Do we need this file?

File: test/unit/plugin/v2/back_compat_test.rb
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/back_compat_test.rb]: Any reason this is commented?

File: test/unit/plugin/v2/api_cli_test.rb
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/api_cli_test.rb]: Do we need this file?

File: docs/dev/plugins.md
Comment (jerryaldrichiii) [inline docs/dev/plugins.md]: Could we capitalize `InSpec`?

File: docs/dev/plugins.md
Comment (jerryaldrichiii) [inline docs/dev/plugins.md]: Also, here `InSpec`?

File: docs/dev/plugins.md
Comment (jerryaldrichiii) [inline docs/dev/plugins.md]: I'm gonna refrain from highlighting the InSpec capitalization thing after this one :smile:

File: lib/inspec/dependencies/requirement.rb
Comment (jerryaldrichiii) [inline lib/inspec/dependencies/requirement.rb:3]: Huh...not sure how removing this doesn't break anything.

File: lib/inspec/plugin/v2/plugin_base.rb
Comment (jerryaldrichiii) [inline lib/inspec/plugin/v2/plugin_base.rb]: Weird...thought RuboCop would catch this empty line at top of file.

File: lib/inspec/plugin/v2/plugin_base.rb
Comment (jerryaldrichiii) [inline lib/inspec/plugin/v2/plugin_base.rb]: Needs to be `CliPlugin`

File: test/functional/plugins_test.rb
Comment (jerryaldrichiii) [inline test/functional/plugins_test.rb]: Are these tests going to be added before merge? Should we add them but `skip`?

File: test/unit/plugin/v2/api_cli_test.rb
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/api_cli_test.rb]: Do we need this file? Should we add the tests and use `skip`?

File: test/unit/plugin/v2/installer_test.rb
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/installer_test.rb]: Same as above, do we need this file?

File: test/unit/plugin/v2/loader_test.rb
Comment (jerryaldrichiii) [inline test/unit/plugin/v2/loader_test.rb]: Should we remove these bits? Add them but use skip?

**Review by jerryaldrichiii — CHANGES_REQUESTED** [review]
As per usual, only minor feedback. Mainly capitalization of InSpec and the amount of TODOs/commented out things.

There are also some empty line issues in the tests...but they're not linted so I can't really complain :sob:

**Comment by clintoncwolfe** [comment]
Adding DNM while I sort this out

File: docs/dev/plugins.md
Comment (clintoncwolfe) [reply docs/dev/plugins.md]: You really can just make one comment, along the lines of "Throughout this PR...."

File: lib/inspec/dependencies/requirement.rb
Comment (clintoncwolfe) [reply lib/inspec/dependencies/requirement.rb:3]: Trace the requires.  

1. inspec/profile.rb requires inspec/dependencies/dependency_set
2. inspec/dependencies/dependency_set requires inspec/dependencies/requirement
3. inspec/dependencies/requirement requires inspec/dependencies/dependency_set, triggering a circular dependency warning.  I removed this require to break the loop.

File: test/functional/plugins_test.rb
Comment (clintoncwolfe) [reply test/functional/plugins_test.rb]: It was just my way of keep notes prior to a branching strategy bing determined.  I'll remove them, as these (implemented) tests will come in on another branch.

File: test/unit/plugin/v2/api_cli_test.rb
Comment (clintoncwolfe) [reply test/unit/plugin/v2/api_cli_test.rb]: It was just my way of keep notes prior to a branching strategy bing determined. I'll remove them, as these (implemented) tests will come in on another branch.

File: test/unit/plugin/v2/installer_test.rb
Comment (clintoncwolfe) [reply test/unit/plugin/v2/installer_test.rb]: It was just my way of keep notes prior to a branching strategy bing determined. I'll remove them, as these (implemented) tests will come in on another branch.

File: test/unit/plugin/v2/loader_test.rb
Comment (clintoncwolfe) [reply test/unit/plugin/v2/loader_test.rb]: These have become irrelevant, I've removed them - thanks

File: lib/inspec/dependencies/requirement.rb
Comment (jerryaldrichiii) [reply lib/inspec/dependencies/requirement.rb:3]: Ah! Great catch.

File: docs/dev/plugins.md
Comment (jerryaldrichiii) [inline docs/dev/plugins.md:39]: Missed one `InSpec`, I won't hold up a merge over it though.

**Review by jerryaldrichiii — APPROVED** [review]
Looks good, only one missing `InSpec` cap issue. I won't hold up the merge over it though.

**Pull request merged**
